src/engine-actions/translate-swap.js

```diff
@@ -1,220 +1,74 @@
- * @param {Object} options
- * @param {Object} options.logger
- * @param {LndClient} options.client
- * @returns {Promise<TranslateSwapOutcome>}
- * @throws {Error} unknow lnd error
- * @throws {Error} lnd closed stream to send to route
- * @throws {Error} stream failed to finish
- */
-function sendToRouteSync (paymentHash, routes, { logger, client }) {
-  return new Promise((resolve, reject) => {
-    try {
-      logger.debug('Setting up stream to sendToRoute', { paymentHash })
-      const stream = sendToRoute({ client })
-
-      // keep local state of whether we (as opposed to LND) cancelled so
-      // that we can handle a CANCELLED error message appropriately
-      let clientCancel = false
-
-      const errorListener = (err = new Error('Unknown LND Error')) => {
-        // if we cancelled the request, we treat a CANCELLED code
-        // as a cleanup and not a true error
-        if (clientCancel && err.code && err.code === CANCELLED_CODE) {
-          logger.debug('Received CANCELLED after client cancelled, cleaning up')
-          finish()
-        } else {
-          logger.error(`Error from sendToRoute stream`, err)
-          finish(err)
-        }
-      }
-
-      const endListener = () => {
-        logger.error(`LND closed sendToRoute stream before returning a value`)
-        return finish(new Error(`LND closed stream to send to route`))
-      }
-
-      const dataListener = (response) => {
-        const { paymentError, paymentPreimage } = response
-        logger.debug('Received sendToRoute response from LND', { paymentError, paymentPreimage })
-
-        // cancel will trigger cleanup through our errorListener
-        clientCancel = true
-        stream.cancel()
-
-        // TODO: do we need to ensure that this preimage matches the hash that we sent?
-        resolve({ paymentError, paymentPreimage })
-      }
-
-      // Helper to make sure we tear down our listeners
-      const finish = (err, response) => {
-        stream.removeListener('error', errorListener)
-        stream.removeListener('end', endListener)
-        stream.removeListener('data', dataListener)
-
-        if (err) {
-          return reject(err)
-        }
-      }
-
-      stream.on('error', errorListener)
-      stream.on('end', endListener)
-      stream.on('data', dataListener)
-
-      logger.debug('Writing to sendToRoute stream', { paymentHash })
-      stream.write({ paymentHash, routes })
-    } catch (e) {
-      reject(e)
-    }
-  })
-}
+const DEFAULT_FEE_LIMIT = 0
 
 /**
  * @typedef {Object} TranslateSwapOutcome
- * @property {String} paymentPreimage Base64 string of the preimage for the swapHash
- * @property {String} permanentError Error encountered that is permanent, and safe to cancel the upstream HTLC
+ * @property {String} paymentPreimage Base64 string of the preimage for the
+ *                                    swapHash
+ * @property {String} permanentError Error encountered that is permanent, and
```

Comment:
Would it be possible to change this to an exception and just return the preimage? Using an object property to indicate errors is a little unconventional and makes it easier to make silent bugs when the caller forgets to check for the property.

Reply:
Yes, we could potentially have a specific error type, like we do in https://github.com/sparkswap/lnd-engine/pull/234.

Reply:
One issue we may have with these custom error types is how they are used when we have multiple engines. Each engine will have a different custom error type, so type comparison won't work.

What do you think @clark800 ?

Reply:
One idea would be to expose them on the instance of the engine, but that seems pretty nasty.

Reply:
I think exposing the error types through an export seems reasonable, and they could all use the same name, like `PermanentError`, so that they can be imported easily. Another option is to use `error.constructor.name` to test which would avoid needing to export them, but I don't see much downside in exporting.